### PR TITLE
[spirv] Legalization: associated counters in implicit objects

### PR DESCRIPTION
**Pull request by antiagainst**
An implicit object is translated into the first argument to the
method call. If the implicit object is of struct type and its
fields have associated counters, we need to assign its associated
counters accordingly like other normal arguments.

For each such method, we generate associated counters for its
implicit object. At a call site, we assign the associated counters
from the real object to the ones associated with the method implict
object.

### CODE REVIEW

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 1]
```diff
@@ -2822,11 +2829,9 @@ bool SPIRVEmitter::tryToAssignCounterVar(const DeclaratorDecl *dstDecl,
   }
 
   // Handle AssocCounter#3
-  const auto *dstFields = declIdMapper.getCounterVarFields(dstDecl);
   llvm::SmallVector<uint32_t, 4> srcIndices;
-  const auto *srcDecl = getReferencedDef(
-      collectArrayStructIndices(srcExpr, &srcIndices, /*rawIndex=*/true));
-  const auto *srcFields = declIdMapper.getCounterVarFields(srcDecl);
+  const auto *dstFields = declIdMapper.getCounterVarFields(dstDecl);
+  const auto *srcFields = getIntermediateACSBufferCounter(srcExpr, &srcIndices);
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: +1. This function makes things cleaner.

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 2]
```diff
@@ -2889,13 +2890,47 @@ SPIRVEmitter::getFinalACSBufferCounter(const Expr *expr) {
...
-          collectArrayStructIndices(expr, &indices, /*rawIndex=*/true)))
-    return declIdMapper.getCounterIdAliasPair(decl, &indices);
+
+  const auto *base =
+      collectArrayStructIndices(expr, &indices, /*rawIndex=*/true);
+  const auto *decl =
+      (base && isa<CXXThisExpr>(base))
+          ? getOrCreateDeclForMethodObject(cast<CXXMethodDecl>(curFunction))
+          : getReferencedDef(base);
+  return declIdMapper.getCounterIdAliasPair(decl, &indices);
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: return followed by ``return nullptr``!

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 3]
```diff
@@ -2889,13 +2890,47 @@ SPIRVEmitter::getFinalACSBufferCounter(const Expr *expr) {
...
+SPIRVEmitter::getOrCreateDeclForMethodObject(const CXXMethodDecl *method) {
+  const auto found = thisDecls.find(method);
+  if (found != thisDecls.end())
+    return found->second;
+
+  const std::string name = method->getName().str() + ".this";
+  // Create a new identifier to convey the name
+  auto &identifier = astContext.Idents.get(name);
+
+  return thisDecls[method] = ImplicitParamDecl::Create(
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: nice!

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 2 above)
Comment (antiagainst) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Good catch! Will remove in a following up PR.

**Pull request merged**